include/alpaka/event/EventCpu.hpp

```diff
@@ -77,16 +77,25 @@ namespace alpaka
                     }
 
                     //-----------------------------------------------------------------------------
-                    auto hasBeenReadieadSince(const std::size_t & enqueueCount) noexcept -> bool
+                    auto wait(std::size_t const & enqueueCount) noexcept -> void
                     {
-                        return (m_LastReadyEnqueueCount >= enqueueCount);
+                        std::unique_lock<std::mutex> lk(m_mutex);
+                        assert(enqueueCount <= m_enqueueCount);
+
+                        while(enqueueCount > m_LastReadyEnqueueCount)
+                        {
+                            auto future = m_future;
+                            lk.unlock();
```

Comment:
Shoudln't that the other way around? locking, working, unlocking?

Reply:
No, I think this is correct. At a specific point A in time we want to wait. In this case we lock the event mutex, check if we need to wait at all and copy the current future. Then we unlock the event and wait for the future to finish. This waiting can not be done in locked moder because it might deadlock for multiple reasons but most prominently because the execution of the event itself requires the lock to increase the `m_LastReadyEnqueueCount`.
We waited on a copy of the future because in the meantime the event could have been re-enqueued to a later point in time. This is no problem because after waiting we simply lock the event again and and the while check is false and we correctly waited for the event.

Reply:
You are right, sorry, I mixed up `lock` and `mutex` :man_facepalming: